ui/apps/dashboard/src/app/api/inngest/functions/agents/query-writer.ts

```diff
@@ -0,0 +1,224 @@
+import { createAgent, createTool, openai, type AnyZodType } from '@inngest/agent-kit';
+import { z } from 'zod';
+
+import type { GenerateSqlResult, InsightsAgentState as InsightsState } from './types';
+
+const queryGrammar = `
+  QueryAST = "SELECT" SelectClause "FROM" From ("WHERE" Expression)? ("GROUP" "BY" GroupBy)? ("ORDER" "BY" OrderBy)? ("LIMIT" <number>)? ("OFFSET" <number>)? ";"? .
+  SelectClause = "*" | (AliasedColumnExpression ("," AliasedColumnExpression)*) .
+  AliasedColumnExpression = (FunctionCall | ColumnReference) ("AS" <ident>)? .
+  FunctionCall = <ident> "(" (FunctionArgument ("," FunctionArgument)*)? ")" .
+  FunctionArgument = FunctionCall | Operand .
+  Operand = Summand ("|" "|" Summand)* .
+  Summand = Factor (("+" | "-") Factor)? .
+  Factor = Term (("*" | "/" | "%") Term)? .
+  Term = ValueType | FunctionCall | ColumnReference .
+  ValueType = ("*" | <number> | <string> | ("TRUE" | "FALSE") | "NULL") .
+  ColumnReference = (<ident> ".")? <ident> .
+  From = TableExpression .
+  TableExpression = <ident> ("AS" <ident>)? .
+  Expression = OrCondition ("OR" OrCondition)* .
+  OrCondition = Condition ("AND" Condition)* .
+  Condition = ConditionOperand .
+  ConditionOperand = Operand ConditionRHS? .
+  ConditionRHS = Compare .
+  Compare = ("<>" | "<=" | ">=" | "=" | "<" | ">" | "!=") Operand .
+  GroupBy = Expression ("," Expression)* .
+  OrderBy = OrderExpression ("," OrderExpression)* .
+  OrderExpression = Expression ("ASC" | "DESC")? .
+`;
+
+const queryRules = `
+Inngest Insights allows you to query and analyze your event data using SQL directly within the Inngest platform. Every event sent to Inngest contains valuable information, and Insights gives you the power to extract meaningful patterns and analytics from that data.
+
+Overview
+Insights provides an in-app SQL editor and query interface where you can:
+
+Query event data using familiar SQL syntax
+Save and reuse common queries
+Analyze patterns in your event streams
+Extract business intelligence from your workflows
+Currently, you can query events only. Support for querying function runs will be added in future releases.
+
+Getting Started
+Available Columns
+When querying events, you have access to the following columns:
+
+Column	Type	Description
+id	String	Unique identifier for the event
+name	String	The name/type of the event
+data	JSON	The event payload data - users can send any JSON structure here
+ts	DateTime	Timestamp when the event occurred
```

Comment:
this line (and maybe others) is out of date https://github.com/inngest/website/blob/4ac846e835044c7cdc9974493afcca9b95a60fdd/pages/docs/platform/monitor/insights.mdx?plain=1#L53